go/arrow/compare.go

```diff
@@ -0,0 +1,58 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arrow
+
+import (
+	"reflect"
+)
+
+// TypeEquals checks if two DataType are the same, optionally checking metadata
+// equality for STRUCT types.
+func TypeEquals(left DataType, right DataType, checkMetadata bool) bool {
+	switch {
+	case left == nil || right == nil:
+		return false
```

Comment:
shouldn't we return `left == right` here? if both `left` and `right` are `nil`, they'd logically test equal. right?

Reply:
I agree with your logic but the idea was that two nil types are "undefined" so they're not equal. Like I don't consider two NULL values to be equal but it's a matter of convention. I will update the return statement to your suggestion.

Reply:
ah, so you consider `nil` `DataType`s to behave like `NaN`s?

ok. I must say I am not sure what the arrow specs in this area :)

Reply:
looking at:
https://github.com/apache/arrow/blob/master/cpp/src/arrow/type.cc#L110

it would seem C++ agrees with your original implementation.
it's of course constrained by the fact in C++ the feature is implemented as a method on a type rather than a function taking 2 args (so in C++ there's no possibilty to have both args to be `nullptr_t`.)

oh well...

Reply:
Hum... I was looking at https://github.com/apache/arrow/blob/8ccf48181a42a1a08d2a172dd1c958f82599ad9f/cpp/src/arrow/compare.cc#L1046 and it seems to suggest the opposite 🤔

Reply:
if my C++ isn't completely rusty, you can't have `&left` nor `&right` being `nullptr` (as they're both references.)

Reply:
Oh yes you're right. Mine is completely rusty :)